### PR TITLE
Mccalluc/try to split deps again

### LINKED ISSUE #127: Split widget code off as an optional install
I’m running into a conflicting version constraint problem, basically because PSC is unable to upgrade from Python 3.6 :slightly_frowning_face: . I’m trying to move a bit of the view-conf generation to index-time, but vitessce-python has a dependency on a hypercorn version that isn’t supported on 3.6.

I don’t think we should bend over backward for this… but I also wonder if it would make sense to split out the widget dependencies from the rest of vitessce-python… something like pip install `vitessce-python[widget]`. Or something like that… Beyond hypercorn, I think there are other dependencies that interactive use requires, that just aren’t needed for building viewconfs.

- Original goal: https://github.com/hubmapconsortium/portal-visualization/issues/37
- Blocked PR: https://github.com/hubmapconsortium/search-api/pull/467
- The blocker: https://github.com/hubmapconsortium/gateway/issues/191
- See also: #122

### PR DESCRIPTION
**Pull request by mccalluc**
- Fix #127 
- Replace #147

Reached out to Mark for guidance on the best way forward:

> Thanks for the feedback, and sorry this has been drawn out: I finally got it to install under Python 3.6 😄  … but I needed to take out numpy and generate-tiff-offsets ☹️  . Do you have a sense of the best way forward from here?
> - Try to downgrade these dependencies to a version that is supported under Python 3.6?
> - Or move the imports down inside GenomicProfiles and SnapWrapper and OmeTiffWrapper ?
> - Or a combination?
> - Or give up — Trying to support compatibility with 3.6 might just not be worth it?

Went with moving the imports down... this gets pretty kludgy -- Unless identifying the dependencies will make a future clean up easier, I have misgivings about pushing this in.

### CODE REVIEW

**Comment by mccalluc** [comment]
```
The conflict is caused by:
ERROR: ResolutionImpossible: for help visit https://pip.pypa.io/en/latest/topics/dependency-resolution/#dealing-with-dependency-conflicts
    vitessce[notebook] 1.0.5 depends on vitessce 1.0.5 (from /home/runner/work/vitessce-python/vitessce-python)
    vitessce[testing] 1.0.5 depends on vitessce 1.0.5 (from /home/runner/work/vitessce-python/vitessce-python)
```
Maybe it has some problem resolving this when they're both local? Instead of having multiple extra installs, I could include the `notebook` deps under `docs` and `tests` extras?

**Comment by mccalluc** [comment]
I'm going to keep working on this branch, and would like to identify a subset of the tests that cover conf generation, but this much can be reviewed.

File: .github/workflows/test_python_36.yml [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+          python-version: '3.6'
+      - name: Install vitessce
+        run: pip install -e .
+
+      # For now, just confirm that it can be installed.
+      # 3.6 support may only be needed for the short term, so may not be worth splitting tests.
+
+      # - name: Run unittests
+      #   working-directory: ./tests
+      #   run: python -m unittest
```
Comment (mccalluc) [inline .github/workflows/test_python_36.yml:22]:
```suggestion
      - name: Run unittests
        working-directory: ./tests
        run: python -m unittest
```

File: .github/workflows/test_python_36.yml [hunk 2]
```diff
@@ -0,0 +1,25 @@
+name: Test Python 3.6 (install only)
```
Comment (mccalluc) [inline .github/workflows/test_python_36.yml:1]:
```suggestion
name: Test Python 3.6 (partial)
```

File: setup.py [hunk 3]
```diff
@@ -77,9 +77,29 @@
...
+
+        # These are used in wrapper classes, and are not compatible with Python 3.6:
+        'numpy>=1.21.2',  # Last compatible version is 1.19.5
+        'generate-tiff-offsets>=0.1.7',  # Last compatible version is 2020.9.3
     ]
 }
 
+extras_require['docs'] += extras_require['notebook']
+extras_require['testing'] += extras_require['notebook']
+
```
Comment (mccalluc) [inline setup.py:102]:
```suggestion
```

I was importing things wrong, and this was a work around.

File: vitessce/__init__.py [hunk 4]
```diff
@@ -2,33 +2,47 @@
...
+    )
+    from .export import (
+        export_to_s3,
+        export_to_files,
+    )
+except ModuleNotFoundError as e:
+    from sys import version_info
+    if version_info >= (3, 7):
+        raise e
+    # If version < 3.7, these exports just aren't available
```
Comment (mccalluc) [inline vitessce/__init__.py:44]:
```suggestion
    # If version < 3.7, these exports just aren't available.
    # In the long term, probably best to drop partial support for 3.6, when it's no longer needed.
```

File: vitessce/config.py [hunk 5]
```diff
@@ -1150,6 +1144,7 @@ def export(self, to, *args, **kwargs):
 
             config_dict = vc.export(to="S3")
         """
+        from .export import (export_to_s3, export_to_files)
```
Comment (mccalluc) [inline vitessce/config.py:1147]:
```suggestion
        from .export import (export_to_s3, export_to_files)  # TODO: Move import back to top when this is factored out.
```

File: vitessce/wrappers.py [hunk 6]
```diff
@@ -228,10 +219,17 @@ def make_raster_routes(self, dataset_uid, obj_i):
...
+            from .routes import range_repsonse, JsonRoute
+            from generate_tiff_offsets import get_offsets
+            from starlette.responses import UJSONResponse
+            from starlette.routing import Route
+
             offsets = get_offsets(self._img_path)
 
             async def response_func(req):
                 return UJSONResponse(offsets)
+
```
Comment (mccalluc) [inline vitessce/wrappers.py:232]:
```suggestion
```

**Review by mccalluc — COMMENTED** [review]
Catching my own typos...

**Review by manzt — APPROVED** [review]
This seems like a sufficient solution to keep py3.6 support. If I recall correctly, the 3.6 requirement is coming from the PSC? Otherwise I would say that 3.6 is more trouble than it's worth. (Projects following [NEP 29 ](https://numpy.org/neps/nep-0029-deprecation_policy.html) have already started dropping 3.7. However, I believe Google Colab still is 3.7 so that would be the minimal python version I would target for vitessce-python otherwise)

**Comment by mccalluc** [comment]
>  If I recall correctly, the 3.6 requirement is coming from the PSC?

Yep, that's the motivation. There are portions of the HuBMAP process which are uncontainerized. (Something about access to the filesystem that wouldn't work under Docker?) The host environment there is out of date, and there's no clear timeline to get it fixed.

**Pull request merged**